### PR TITLE
Fix arrays with zero-size dimensions

### PR DESCRIPTION
**Pull request by mwoehlke-kitware**
When converting an array to an Eigen matrix, ignore the strides if any dimension size is 0. If the array is empty, the strides aren't relevant, and especially numpy ≥ 1.23 explicitly sets the strides to 0 in this case. (See numpy/numpy@dd5ab7b11520.)

Update tests to verify that this works, and continues to work. (The updated test would have caught the behavior change with numpy 1.23.) This expands on the tests added by #38, but also reverts the logic change from that PR, as it is no longer needed, and brings the code closer in line with upstream, since #38 was never pushed upstream.

+@EricCousineau-TRI for review, please.

See also https://github.com/pybind/pybind11/pull/4038.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/58)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
We can't merge patches here unless we _also_ open a Drake PR that refers to this PR, to prove that Drake CI passes (not just RobotLocomotion/pybind11 CI).  Please open the Drake PR and mention it here.

**Comment by mwoehlke-kitware** [comment]
Drake CI succeeded for https://github.com/RobotLocomotion/drake/pull/17475. Particularly:
- https://drake-jenkins.csail.mit.edu/job/mac-monterey-clang-bazel-experimental-everything-release/8/
- https://drake-jenkins.csail.mit.edu/job/mac-big-sur-clang-bazel-experimental-everything-release/45/

...which should be using numpy 1.23 and thus would be failing without this change.

Meanwhile, upstream merged a version with (non-functional) changes. I'd rather not wait on all the CI to re-run; is it going to be a problem for whoever ends up merging upstream to remember to just take their version, or should I tweak this to more closely match the versions that's now upstream?

**Comment by jwnimmer-tri** [comment]
Definitely our patch should be as similar to upstream as possible.  If the timeline means foregoing another spin of Drake CI, so be it.

**Review by jwnimmer-tri — APPROVED** [review]
:lgtm: rubber stamp, pending any re-sync with the upstream PR.

**Comment by mwoehlke-kitware** [comment]
> Definitely our patch should be as similar to upstream as possible.

Okay, done. It should be easy enough to verify in Reviewable that there is no functional change, just aesthetic changes. Once this lands, I'll update the Drake PR to point to the post-merge SHA.

**Comment by mwoehlke-kitware** [comment]
...or it *would* have been if Reviewable had kept the previous version. Since it looks like it didn't, here's that diff:
```diff
-        if (rows == 0 || cols == 0)
-            return !negativestrides;
-        return
-            !negativestrides &&
-            (props::inner_stride == Eigen::Dynamic || props::inner_stride == stride.inner() ||
-                (EigenRowMajor ? cols : rows) == 1) &&
-            (props::outer_stride == Eigen::Dynamic || props::outer_stride == stride.outer() ||
-                (EigenRowMajor ? rows : cols) == 1);
+        if (negativestrides) {
+            return false;
+        }
+        if (rows == 0 || cols == 0) {
+            return true;
+        }
+        return (props::inner_stride == Eigen::Dynamic || props::inner_stride == stride.inner()
+                || (EigenRowMajor ? cols : rows) == 1)
+               && (props::outer_stride == Eigen::Dynamic || props::outer_stride == stride.outer()
+                   || (EigenRowMajor ? rows : cols) == 1);
```

Note also that everything starting with `(props::inner_stride` differs only in whitespace.

**Pull request merged**
